File: typescript/packages/x402/src/signers/evmSigner.ts
/**
 * EvmClientSigner - EVM client signer for x402 protocol
 */

import {
  createPublicClient,
  http,
  type PublicClient,
  type Hex,
  parseAbi,
  encodeFunctionData,
  type Chain,
} from 'viem';
import { mainnet, sepolia, bsc, bscTestnet } from 'viem/chains';
import type { ClientSigner } from '../client/x402Client.js';
import {
  getPaymentPermitAddress,
  resolveRpcUrl,
  InsufficientAllowanceError,
  UnsupportedNetworkError,
} from '../index.js';

import type { AgentWallet } from './signer.js';

const ERC20_ABI = parseAbi([
  'function allowance(address owner, address spender) view returns (uint256)',
  'function approve(address spender, uint256 amount) returns (bool)',
  'function balanceOf(address account) view returns (uint256)',
]);

/**
 * EVM client signer implementation.
 *
 * Accepts any wallet conforming to the AgentWallet interface.
 * The signer is agnostic about how the wallet was created
 * (private key, hosted, etc.).
 */
export class EvmClientSigner implements ClientSigner {
  private wallet: AgentWallet;
  private _address: string;
  private publicClients: Map<number, PublicClient> = new Map();

  /**
   * Create signer from a wallet and its pre-resolved address.
   *
   * Prefer the async factory `EvmClientSigner.create()` which resolves
   * the address automatically.
   */
  constructor(wallet: AgentWallet, address: string) {
    this.wallet = wallet;
    this._address = address;
  }

  /** Async factory: resolve address from wallet and create signer. */
  static async create(wallet: AgentWallet): Promise<EvmClientSigner> {
    const address = await wallet.getAddress();
    return new EvmClientSigner(wallet, address);
  }

  getAddress(): string {
    return this._address;
  }

  getEvmAddress(): Hex {
    return this._address as Hex;
  }

  async signMessage(message: Uint8Array): Promise<string> {
    return this.wallet.signMessage(message);
  }

  async signTypedData(
    domain: Record<string, unknown>,
    types: Record<string, unknown>,
    message: Record<string, unknown>,
    primaryType: string
  ): Promise<string> {
    const fullData = {
      types: { EIP712Domain: [], ...types },
      domain,
      primaryType,
      message,
    };

    return this.wallet.signTypedData(fullData);
  }

  async checkBalance(token: string, network: string, address?: string): Promise<bigint> {
    const chainId = this.parseNetworkToChainId(network);
    const client = this.getPublicClient(chainId, network);
    try {
      return await client.readContract({
        address: token as Hex,
        abi: ERC20_ABI,
        functionName: 'balanceOf',
        args: [(address ?? this._address) as Hex],
      });
    } catch (error) {
      console.error(
        `[EvmClientSigner] checkBalance failed for ${token} on ${network}:`,
        error,
      );
      return 0n;
    }
  }

  async checkAllowance(
    token: string,
    _amount: bigint,
    network: string,
  ): Promise<bigint> {
    const chainId = this.parseNetworkToChainId(network);
    const client = this.getPublicClient(chainId, network);
    const spender = getPaymentPermitAddress(network) as Hex;

    try {
      return await client.readContract({
        address: token as Hex,
        abi: ERC20_ABI,
        functionName: 'allowance',
        args: [this._address as Hex, spender],
      });
    } catch (error) {
      console.error(
        `[EvmClientSigner] checkAllowance failed for ${token} on ${network}:`,
        error,
      );
      return 0n;
    }
  }

  async ensureAllowance(
    token: string,
    amount: bigint,
    network: string,
    mode: 'auto' | 'interactive' | 'skip' = 'auto',
  ): Promise<boolean> {
    if (mode === 'skip') return true;

    const currentAllowance = await this.checkAllowance(token, amount, network);
    if (currentAllowance >= amount) return true;

    if (mode === 'interactive') {
      throw new InsufficientAllowanceError('Interactive approval required');
    }

    const chainId = this.parseNetworkToChainId(network);
    const client = this.getPublicClient(chainId, network);
    const spender = getPaymentPermitAddress(network) as Hex;

    try {
      // Build approve calldata
      const data = encodeFunctionData({
        abi: ERC20_ABI,
        functionName: 'approve',
        args: [spender, BigInt(2) ** BigInt(256) - BigInt(1)],
      });

      // Prepare transaction via public client
      const nonce = await client.getTransactionCount({ address: this._address as Hex });
      const gasPrice = await client.getGasPrice();
      const gas = await client.estimateGas({
        account: this._address as Hex,
        to: token as Hex,
        data,
      });

      const tx = {
        from: this._address,
        to: token,
        data,
        nonce,
        gas: Number(gas),
        gasPrice: Number(gasPrice),
        chainId,
      };

      // Sign via wallet and broadcast
      const signedTxHex = await this.wallet.signTransaction(tx);
      const hash = await client.sendRawTransaction({
        serializedTransaction: `0x${signedTxHex.replace(/^0x/, '')}` as Hex,
      });

      const receipt = await client.waitForTransactionReceipt({ hash });

      const success = receipt.status === 'success';
      if (success) {
        console.info(
          `[EvmClientSigner] ERC20 approval confirmed for ${token}, tx: ${hash}`,
        );
      }
      return success;
    } catch (error) {
      console.error(
        `[EvmClientSigner] ERC20 approval failed for ${token}:`,
        error,
      );
      throw new InsufficientAllowanceError(
        `ERC20 approval transaction failed for ${token}: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
  }

  private getPublicClient(chainId: number, network: string): PublicClient {
    let client = this.publicClients.get(chainId);
    if (!client) {
      const rpcUrl = resolveRpcUrl(network);
      client = createPublicClient({
        chain: this.getChain(chainId),
        transport: http(rpcUrl),
      });
      this.publicClients.set(chainId, client);
    }
    return client;
  }

  private getChain(chainId: number): Chain {
    const chains: Record<number, Chain> = {
      1: mainnet,
      11155111: sepolia,
      56: bsc,
      97: bscTestnet,
    };

    const chain = chains[chainId];
    if (!chain) {
      throw new UnsupportedNetworkError(`Unsupported EVM chain ID: ${chainId}`);
    }
    return chain;
  }

  private parseNetworkToChainId(network: string): number {
    if (!network.startsWith('eip155:')) {
      throw new UnsupportedNetworkError(
        `Invalid EVM network format: ${network}`,
      );
    }
    const chainId = parseInt(network.split(':')[1], 10);
    if (isNaN(chainId)) {
      throw new UnsupportedNetworkError(`Invalid EVM chain ID in: ${network}`);
    }
    return chainId;
  }
}

File: legacy/python/x402/src/bankofai/x402/signers/client/evm_signer.py
"""
EvmClientSigner - EVM client signer implementation

Accepts any wallet object that exposes the agent-wallet Wallet interface
(get_address, sign_message, sign_typed_data, sign_transaction).
The signer is agnostic about how the wallet was created (private key, hosted, etc.).
"""

import logging
from typing import Any

from bankofai.x402.abi import ERC20_ABI
from bankofai.x402.config import NetworkConfig
from bankofai.x402.exceptions import InsufficientAllowanceError, SignatureCreationError
from bankofai.x402.signers.client.base import ClientSigner
from bankofai.x402.signers.utils import resolve_provider_uri
from bankofai.x402.utils.address import checksum_evm_address

logger = logging.getLogger(__name__)


class EvmClientSigner(ClientSigner):
    """EVM client signer implementation using web3.py"""

    def __init__(self, wallet: Any) -> None:
        """Create signer from a wallet.

        Prefer the async factory ``create()``.

        Args:
            wallet: Any object implementing the Wallet interface
                    (get_address, sign_message, sign_typed_data, sign_transaction).
        """
        self._wallet = wallet
        self._address: str | None = None
        self._async_web3_clients: dict[str, Any] = {}
        logger.debug("EvmClientSigner initialized")

    @classmethod
    async def create(cls) -> "EvmClientSigner":
        """Async factory: resolve active agent wallet and create signer."""
        from agent_wallet import resolve_wallet_provider

        provider = resolve_wallet_provider(network="eip155")
        wallet = await provider.get_active_wallet()
        signer = cls(wallet)
        signer.set_address(await wallet.get_address())
        return signer

    def get_address(self) -> str:
        if not self._address:
            raise ValueError("Signer address has not been initialized")
        return self._address

    def set_address(self, address: str) -> None:
        self._address = address

    def _ensure_async_web3_client(self, network: str) -> Any:
        """Lazy initialize async web3 client for the given network."""
        if network not in self._async_web3_clients:
            from web3 import AsyncHTTPProvider, AsyncWeb3
            from web3.middleware import ExtraDataToPOAMiddleware

            provider_uri = resolve_provider_uri(network)
            w3 = AsyncWeb3(AsyncHTTPProvider(provider_uri))
            w3.middleware_onion.inject(ExtraDataToPOAMiddleware, layer=0)
            self._async_web3_clients[network] = w3

        return self._async_web3_clients[network]

    async def sign_message(self, message: bytes) -> str:
        """Sign raw message using ECDSA (EIP-191)"""
        try:
            return await self._wallet.sign_message(message)
        except Exception as e:
            raise SignatureCreationError(f"Failed to sign message: {e}")

    async def sign_typed_data(
        self,
        domain: dict[str, Any],
        types: dict[str, Any],
        message: dict[str, Any],
        primary_type: str,
    ) -> str:
        """Sign EIP-712 typed data (Pure Passthrough)."""
        try:
            full_data = {
                "types": types,
                "domain": domain,
                "primaryType": primary_type,
                "message": message,
            }

            return await self._wallet.sign_typed_data(full_data)
        except Exception as e:
            raise SignatureCreationError(f"Failed to sign typed data: {e}")

    async def check_balance(self, token: str, network: str, address: str | None = None) -> int:
        """Check ERC20 token balance"""
        try:
            w3 = self._ensure_async_web3_client(network)
            if not w3:
                return 0

            target_address = checksum_evm_address(address or self._address)
            token_address = checksum_evm_address(token)
            contract = w3.eth.contract(address=token_address, abi=ERC20_ABI)
            return await contract.functions.balanceOf(target_address).call()
        except (ImportError, ModuleNotFoundError):
            logger.warning("web3 not available, returning 0 balance")
            return 0
        except Exception as e:
            logger.error("Failed to check ERC20 balance: %s", e)
            raise

    async def check_allowance(self, token: str, amount: int, network: str) -> int:
        """Check ERC20 allowance"""
        try:
            spender = checksum_evm_address(self._get_spender_address(network))
            w3 = self._ensure_async_web3_client(network)
            if not spender or not w3:
                return 0

            token_address = checksum_evm_address(token)
            owner_address = checksum_evm_address(self._address)
            contract = w3.eth.contract(address=token_address, abi=ERC20_ABI)
            return await contract.functions.allowance(owner_address, spender).call()
        except (ImportError, ModuleNotFoundError):
            logger.warning("web3 not available, returning 0 allowance")
            return 0
        except Exception as e:
            logger.error("Failed to check ERC20 allowance: %s", e)
            raise

    async def ensure_allowance(
        self,
        token: str,
        amount: int,
        network: str,
        mode: str = "auto",
    ) -> bool:
        """Ensure allowance is sufficient for the spender"""
        if mode == "skip":
            return True

        current = await self.check_allowance(token, amount, network)
        if current >= amount:
            return True

        if mode == "interactive":
            raise InsufficientAllowanceError("Interactive approval required")

        try:
            w3 = self._ensure_async_web3_client(network)
            if not w3:
                raise InsufficientAllowanceError("Web3 provider not configured")

            spender = checksum_evm_address(self._get_spender_address(network))
            token_address = checksum_evm_address(token)
            from_address = checksum_evm_address(self._address)
            contract = w3.eth.contract(address=token_address, abi=ERC20_ABI)

            tx = await contract.functions.approve(spender, 2**256 - 1).build_transaction(
                {
                    "from": from_address,
                    "nonce": await w3.eth.get_transaction_count(from_address),
                    "chainId": await w3.eth.chain_id,
                }
            )

            signed_tx_hex = await self._wallet.sign_transaction(tx)
            tx_hash = await w3.eth.send_raw_transaction(bytes.fromhex(signed_tx_hex))
            receipt = await w3.eth.wait_for_transaction_receipt(tx_hash)

            success = receipt.status == 1
            if success:
                logger.info(
                    "ERC20 approval successful",
                    extra={"token": token, "tx_hash": tx_hash.hex()},
                )
            return success
        except (ImportError, ModuleNotFoundError):
            raise InsufficientAllowanceError("web3 not available for approval")
        except Exception as e:
            raise InsufficientAllowanceError(f"ERC20 approval transaction failed: {e}")

    def _get_spender_address(self, network: str) -> str:
        """Get payment permit contract address (spender)"""
        return NetworkConfig.get_payment_permit_address(network)
